File: processor/src/main/java/io/shreyash/rush/blocks/DesignerProperty.java
package io.shreyash.rush.blocks;

import javax.annotation.processing.Messager;
import javax.lang.model.element.Element;
import javax.lang.model.element.ExecutableElement;
import javax.tools.Diagnostic;

import shaded.org.json.JSONArray;
import shaded.org.json.JSONObject;

public class DesignerProperty extends Block {
  private final ExecutableElement element;
  private final Messager messager;

  private final BlockStore store = BlockStore.getInstance();

  public DesignerProperty(Element element, Messager messager) {
    super(element);
    this.element = (ExecutableElement) element;
    this.messager = messager;
    validate();
  }

  @Override
  String description() {
    return null;
  }

  @Override
  void validate() {
    // Check if the corresponding setter simple property exists.
    final boolean setterExist = store.getAllProperties().stream()
        .anyMatch(el -> el.name().equals(this.name()));

    if (!setterExist) {
      messager.printMessage(Diagnostic.Kind.ERROR,
          "Unable to find corresponding @SimpleProperty annotation for designer property '"
              + this.name() + "'.");
    }
  }

  /**
   * JSON structure:
   * {
   *  "alwaysSend": "false",
   * 	"defaultValue": "Bar",
   * 	"name": "Foo",
   * 	"editorArgs": ["Bar", "Baz"],
   * 	"editorType": "text"
   * }
   */
  @Override
  public JSONObject asJsonObject() {
    final JSONObject propJson = new JSONObject();
    final com.google.appinventor.components.annotations.DesignerProperty annotation
        = this.element.getAnnotation(com.google.appinventor.components.annotations.DesignerProperty.class);

    propJson.put("name", this.name());
    propJson.put("editorType", annotation.editorType());
    propJson.put("alwaysSend", Boolean.toString(annotation.alwaysSend()));
    propJson.put("defaultValue", annotation.defaultValue());

    final JSONArray args = new JSONArray();
    for (final String arg : annotation.editorArgs()) {
      args.put(arg);
    }
    propJson.put("editorArgs", args);

    return propJson;
  }
}

File: processor/src/main/java/io/shreyash/rush/block/DesignerProperty.kt
package io.shreyash.rush.block

import com.google.appinventor.components.annotations.DesignerProperty
import javax.annotation.processing.Messager
import javax.lang.model.element.Element
import javax.lang.model.element.ExecutableElement
import javax.tools.Diagnostic
import io.shreyash.rush.BlockStore
import shaded.org.json.JSONObject

class DesignerProperty(element: Element, private val messager: Messager) : Block(element) {
    private val store = BlockStore.instance
    private val element = element as ExecutableElement

    init {
        runChecks()
    }

    override fun description(): Nothing? = null

    override fun runChecks() {
        // Check if the corresponding setter simple property exists.
        val setterExist = store.properties.any { it.name() == name() }
        if (!setterExist) {
            messager.printMessage(
                Diagnostic.Kind.ERROR,
                "Unable to find corresponding @SimpleProperty annotation for designer property '"
                        + name() + "'."
            )
        }
    }

    /**
     * @return JSON representation of this designer property.
     * {
     *  "alwaysSend": "false",
     *  "defaultValue": "Bar",
     *  "name": "Foo",
     *  "editorArgs": ["Bar", "Baz"],
     *  "editorType": "text"
     * }
     */
    override fun asJsonObject(): JSONObject {
        val annotation = this.element.getAnnotation(DesignerProperty::class.java)
        return JSONObject()
            .put("name", name())
            .put("editorType", annotation.editorType)
            .put("editorArgs", annotation.editorArgs)
            .put("defaultValue", annotation.defaultValue)
            .put("alwaysSend", annotation.alwaysSend.toString())
    }
}
